debian/ltsp-rbd.initramfs-hook

```diff
@@ -0,0 +1,37 @@
+#!/bin/sh
+
+# This file is part of LTSP, https://ltsp.github.io
+# Copyright 2019 the LTSP team, see AUTHORS
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+# Inject LTSP code under initramfs-tools
+
+# Hook scripts (from man initramfs-tools(8))
+# These  are  used  when an initramfs image is created and not included in the
+# image itself. They can however cause files to be included in the image.  Hook
+# scripts are executed under errexit.  Thus  a hook script can abort the
+# mkinitramfs build on possible errors (exitcode != 0).
+
+# Notes:
+# Use functions and local variables, to avoid namespace pollution
+
+# Initramfs hook for ltsp
+
+PREREQ=""
+
+prereqs()
+{
+    echo "$PREREQ"
+}
+
+case "$1" in
+    prereqs)
+        prereqs
+        exit 0
+        ;;
+esac
+
+. /usr/share/initramfs-tools/hook-functions
+
+modules="rbd aes cbc"
+manual_add_modules $modules
```

Comment:
If RBD is to be part of upstream LTSP, a single hook should be enough; there shouldn't be an additional hook file.
The modules"rbd aes cbc" should come from an INITRAMFS_MODULES ltsp.conf variable.

Reply:
Hi @alkisg,

I am a fan of separation of concern. It keeps code clean, readable, understandable and more easy to maintain. I see no reason to merge everything into one file and no apparent benefit.

True, especially the hook file is pretty darn empty and it may seem obvious that it should be merged or INITRAMFS_MODULES should be used, but the logical conclusion would be, to merge the rbd script file as well. And that certainly does not make sense. So we should refrain from from doing so for the hook either, even it if appears to be so easy and practical.

The hook and the script are a unity, which I believe should not be broken up for just the benefit, that there is one file less. If somebody wants to research how this RBD feature actually works, scripts and hooks dirs are the most obvious places to search for the functionality, so there it should reside. To go looking in ltsp.conf is rather far off, if you approach this with existing initramfs knowledge but are new to LTSP. 

I'd rather follow the initramfs way to do things here than force people to spend time figuring out our new approach.

Also, as said, I believe code should invite to be reused. Like I reused this existing rbd initramfs code. And having these files well names and placed in hooks and scripts dirs facilitates that best.

Reply:
Adding the modules increases the initramfs size.
So when other people start implementing hooks for iscsi, nbd (which will also need nbd-client), encrypted disks etc, you'll end up with a huge initramfs that people won't want to use.

This is the main reason that initramfs-tools doesn't include everything; it needs to keep some things optional.

Your code doesn't make RBD optional, it forces these modules to everyone's initramfs.

If you study the initramfs-tools and dracut code, you'll see that they do have provisions to make things optional, similar to ltsp.conf (e.g. /etc/initramfs-tools/modules).

Reply:
@alkisg, 
I understand where you are coming from.

But, initrd.img as is in Ubuntu is already by far the largest thing we are loading to boot - 70M. The 4,6K of scripts/rbd don't really mean anything.

NFS is included by in initramfs-tools by default also. And no matter what you do in LTSP, I won't be able to make it optional if I am not using it, except if I lay hands on the file supplied by initramfs-tools.

However, RBD module loading and inclusion in the initrd could be made optional, if I'd include some magic, which looks at the CMDLINE_BOOT_METHOD in ltsp.conf and checks if boot=rbd is set.

Here is the exact accounting for the space used in initrd for rbd functionality

My current initrd including rbd:
elias@i7:~/src/debian$ ls -lh /srv/tftp/ltsp/ltsp-01/
insgesamt 78M
-rw-r--r-- 1 root root  70M Okt 23 02:15 initrd.img
-rw-r--r-- 1 root root 8,5M Okt 23 02:15 vmlinuz

--------------------------
sum = 714392 Bytes = 697K
--------------------------
-rw-r--r-- 1 root root 528913 Okt  1 00:23 /lib/modules/5.0.0-31-lowlatency/kernel/net/ceph/libceph.ko
-rw-r--r-- 1 root root 155673 Okt  1 00:23 /lib/modules/5.0.0-31-lowlatency/kernel/drivers/block/rbd.ko
-rw-r--r-- 1 root root  24321 Okt  1 00:23 /lib/modules/5.0.0-31-lowlatency/kernel/arch/x86/crypto/aes-x86_64.ko
-rw-r--r-- 1 root root   4640 Okt 22 22:55 /usr/share/initramfs-tools/scripts/rbd
-rw-r--r-- 1 root root    845 Okt 22 22:55 /usr/share/initramfs-tools/hooks/ltsp-rbd-next